godot-core/src/builtin/quaternion.rs

```diff
@@ -130,70 +138,115 @@ impl Quaternion {
-        inv_factor * self + new_factor * to
-    }
-
-    // pub fn spherical_cubic_interpolate(self, b: Self, pre_a: Self, post_b: Self, weight: real) -> Self {}
-    // TODO: Implement godot's function in Rust
-    /*
-        pub fn spherical_cubic_interpolate_in_time(
-            self,
-            b: Self,
-            pre_a: Self,
-            post_b: Self,
-            weight: real,
-            b_t: real,
-            pre_a_t: real,
-            post_b_t: real,
-        ) -> Self {
-        }
-    */
+        let interpolated = self.as_inner().slerpni(to, weight.as_f64());
+
+        let error = interpolated.is_default() && (!self.is_normalized() || !to.is_normalized());
+        assert!(!error, "Slerpni requires normalized quaternions");
+
+        interpolated
+    }
+
+    /// Performs spherical cubic interpolation.
+    ///
+    /// # Panics
+    /// If any quaternions are not normalized.
+    pub fn spherical_cubic_interpolate(
+        self,
+        b: Self,
+        pre_a: Self,
+        post_b: Self,
+        weight: real,
+    ) -> Self {
+        let interpolated =
+            self.as_inner()
+                .spherical_cubic_interpolate(b, pre_a, post_b, weight.as_f64());
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let error = interpolated.is_default() && !self.are_normalized(b, pre_a, post_b);
+        assert!(
+            !error,
+            "Spherical cubic interpolation requires normalized quaternions"
+        );
+
+        interpolated
+    }
+
+    /// Performs spherical cubic interpolation in time.
+    ///
+    /// # Panics
+    /// If any quaternions are not normalized.
+    #[allow(clippy::too_many_arguments)]
+    pub fn spherical_cubic_interpolate_in_time(
+        self,
+        b: Self,
+        pre_a: Self,
+        post_b: Self,
+        weight: real,
+        b_t: real,
+        pre_a_t: real,
+        post_b_t: real,
+    ) -> Self {
+        let interpolated = self.as_inner().spherical_cubic_interpolate_in_time(
+            b,
+            pre_a,
+            post_b,
+            weight.as_f64(),
+            b_t.as_f64(),
+            pre_a_t.as_f64(),
+            post_b_t.as_f64(),
+        );
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let error = interpolated.is_default() && !self.are_normalized(b, pre_a, post_b);
+        assert!(
+            !error,
+            "Spherical cubic interpolation in time requires normalized quaternions"
+        );
+        interpolated
+    }
 
     #[doc(hidden)]
     pub fn as_inner(&self) -> inner::InnerQuaternion {
         inner::InnerQuaternion::from_outer(self)
     }
+
+    #[doc(hidden)]
+    pub fn is_default(&self) -> bool {
+        self.x == 0.0 && self.y == 0.0 && self.z == 0.0 && self.w == 1.0
```

Comment:
```suggestion
        *self == Self::default()
```

Reply:
I originally had that and went away from it because my gut said "bUt PeRfOrMaNcE!" I'll go back to it because its far easier to understand. :+1:

Reply:
Such concerns should be measured, otherwise simpler code is to be preferred 🙂